src/codegen/compile.ml

```diff
@@ -10384,34 +10393,39 @@ and const_exp_matches_pat env pat exp : bool option =
   assert exp.note.Note.const;
   match exp.it with
   | _ when Ir_utils.is_irrefutable pat -> Some true
+  | LitE _
   | PrimE (TagPrim _, _) ->
      let c, _ = compile_const_exp env VarEnv.empty_ae exp in
-     (try ignore (destruct_const_pat VarEnv.empty_ae pat c); Some true with Invalid_argument _ -> Some false)
+     (try ignore (destruct_const_pat env VarEnv.empty_ae pat c); Some true with Invalid_argument _ -> Some false)
   | _ -> None
 
-and destruct_const_pat ae pat const : VarEnv.t = match pat.it with
+and destruct_const_pat env ae pat const : VarEnv.t = match pat.it with
```

Comment:
Why the `env` parameter? Better not have side effects (adding locals) here, if we end up not doing the constant match anyways. Maybe some function you need to call doesn't need the `env` parameter after all?

Reply:
I was hoping that too. `const_lit_of_lit` needs the `env` only for
``` OCaml
| NullLit       -> Const.Vanilla (Opt.null_vanilla_lit env)
```

Reply:
I'm ok with living with that side effect, the likelihood that this is the only null constant in the program is very high. So maybe leave a comment that this is the only side effect.

Or add a constructor `Const.Null`, and defer `null_vanilla_lit` to where where const values are made manifest? Probably cleaner anyways.

Reply:
Vanilla wouldn't be used for `null` anymore, would it?

Reply:
Yeah, hopefully.

Reply:
Looks like it worked: 84d59c8d0

Reply:
Okay, `test/run/refuted-const-float.mo` checks that `=` returns false, and I have manually checked that `let 3.14 = 3.14` is a `nop`. So this is settled.